### PR TITLE
Externalize Preprepare Validity checker

### PR DESCRIPTION
**Pull request by ranchalp**
Given the extra PBFT instance we have for agreement with the checkpoint certificate, I had to work around some things that otherwise would have been left more general, but I think it is still ok. In particular, the verification of the preprepare for the PBFT instance for the checkpoint certificate is stateful, but I want to leave the signature of the event function general, so I made a factory to create the module dynamically with the relevant state for the checkpoint PBFT module.

The factory can also be used for concurrent treatment across regular PBFT instances, but given at the moment we just have the permissive validity checker we might as well just use one stateless instance (as I propose here). 

This PR 'extends' the commits from #493. The last 3 commits are the ones from this PR only.

### CODE REVIEW

File: protos/ordererpb/prepreparevaliditycheckerpb/prepreparevaliditycheckerpb.proto
Comment (ranchalp) [inline protos/ordererpb/prepreparevaliditycheckerpb/prepreparevaliditycheckerpb.proto:1]: deleted in later commit, This file sneaked in when git add on the folder. Sorry

File: protos/proto_converter_tmp_3690055431/generator.go
Comment (ranchalp) [inline protos/proto_converter_tmp_3690055431/generator.go:1]: deleted in later commit, This file sneaked in when git add on the folder. Sorry

File: pkg/iss/iss.go [hunk 1]
```diff
@@ -221,6 +223,17 @@ func New(
 			checkpointpbtypes.StableCheckpointFromPb(iss.lastStableCheckpoint.Pb()),
 		)
 
+		//Instantiate the stateless PVC module for the permissive validity checker
+		// Instantiate a new PBFT orderer.
```
Comment (ranchalp) [inline pkg/iss/iss.go]: this comment has been removed in a later commit.

File: pkg/orderers/factory.go [hunk 2]
```diff
@@ -58,17 +48,11 @@ func Factory(
...
 
 				// Select validity checker
-				var validityChecker common.ValidityChecker
-				switch ValidityCheckerType(p.ValidityChecker) {
-				case PermissiveValidityChecker:
-					validityChecker = common.NewPermissiveValidityChecker()
-				case CheckpointValidityChecker:
-					validityChecker = common.NewCheckpointValidityChecker(hashImpl, chkpVerifier, segment.Membership, issParams, logger)
 
+				if p.PvcId == "prepreparevaliditycheckerpermissive" { // this must change but that's the scope of a different PR
```
Comment (ranchalp) [inline pkg/orderers/factory.go:52]: Comment removed, but note it.

**Comment by ranchalp** [comment]
Note to reviewer: I will rebase from the checkpoint PR once merged so ignore the commits relevant to that should you review this before I merge :)

**Comment by ranchalp** [comment]
Note to reviewer: I will rebase from the checkpoint PR once merged so ignore the commits relevant to that should you review this before I rebase :)

**Comment by matejpavlovic** [comment]
Thanks for the PR! I'll review it when #493 is merged and this one rebased on top, it will be easier to think about it that way.

**Review by matejpavlovic — CHANGES_REQUESTED** [review]
Nice PR!
I only have one substantial comment (see below), the others are mostly nits.
Also, we'll eventually generalize this evne further, such that the validity checker is not specific to the PBFT protocol, but really just validates proposals (of any protocol). It will then be on the ordering protocol to figure out how to talk to the verifier.
But this is already going in that direction. Nice job!

(Again, I fixed some formatting and typo-level things, I'm pushing them directly.)

**Comment by ranchalp** [comment]
>  Also, we'll eventually generalize this evne further, such that the validity checker is not specific to the PBFT protocol, but really just validates proposals (of any protocol). It will then be on the ordering protocol to figure out how to talk to the verifier.

Actually, the real generalization should be to a generic validity checker module (and factory), that takes generic data as bytes (or repeated bytes?) and has two events, check and checked (or analogous names). I thought about doing this for both the preprepare validity checker (the one in this PR) and the checkpoint validity checker ( #493 ), but in benefit of cleanliness using the benefits of the code generator instead of serializing/deserializing ourselves, I opted for this option. I definitely think once we resolve issue #433 we should opt for the generic validity checker module (and factory). @matejpavlovic

File: protos/ordererpb/ordererpb.proto [hunk 3]
```diff
@@ -54,5 +54,5 @@ message PBFTModule{
   PBFTSegment segment          = 1;
   string      availability_id  = 2;
   uint64      epoch            = 3;
-  uint64      validity_checker = 4;
+  string      pvc_id           = 4;
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply protos/ordererpb/ordererpb.proto]: I initially called it pvc (as the PR was submitted before merging the #493 ). Once we settled on a different name for #493 , I went here to change all instances of pvc to ppv. Seems like I missed this one. Resolved locally will submit with other comments.

[the remark this answers is not in the document]
Comment (ranchalp) [reply protos/ordererpb/ordererpb.proto]: Oh In fact I did change it, this is just an old commit in the same PR.

File: protos/ordererpb/pprepvalidatorpb/pprepvalidatorpb.proto [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+// Events
+// ============================================================
+
+message Event {
+  option (mir.event_class) = true;
+
+  oneof type {
+    option (mir.event_type) = true;
+
+    Validatepreprepare validate_preprepare   = 1;
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply protos/ordererpb/pprepvalidatorpb/pprepvalidatorpb.proto]: Locally changed for simplicity (also this is on an outdated commit so I cannot even commit the suggestion).

File: protos/ordererpb/pprepvalidatorpb/pprepvalidatorpb.proto [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    preprepareValidated preprepare_validated = 2;
+  }
+}
+
+message Validatepreprepare {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply protos/ordererpb/pprepvalidatorpb/pprepvalidatorpb.proto:29]: Locally changed for simplicity with commits

File: pkg/orderers/common/prepreparevaliditychecker/api.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package prepreparevaliditychecker
+
+import pbftpbtypes "github.com/filecoin-project/mir/pkg/pb/pbftpb/types"
+
+// PreprepareValidityChecker is the interface of an external checker of validity of proposed data.
+type PreprepareValidityChecker interface {
+
+	// Verify verifies the validity of the checkpoint
+	// Returns nil on success (i.e., if the given signature is valid) and a non-nil error otherwise.
+	Check(preprepare *pbftpbtypes.Preprepare) error
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/common/prepreparevaliditychecker/api.go]: Good catch. Changed

File: pkg/orderers/common/prepreparevaliditychecker/prepreparevaliditychecker.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	pvcpbtypes "github.com/filecoin-project/mir/pkg/pb/ordererpb/prepreparevaliditycheckerpb/types"
+	pbftpbtypes "github.com/filecoin-project/mir/pkg/pb/pbftpb/types"
+	trantorpbtypes "github.com/filecoin-project/mir/pkg/pb/trantorpb/types"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+// ModuleConfig sets the module ids.
+type ModuleConfig struct {
+	Self t.ModuleID
+	CVC  t.ModuleID
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/common/prepreparevaliditychecker/prepreparevaliditychecker.go]: This was changed already in a later commit once #493 was merged.

File: pkg/orderers/common/prepreparevaliditychecker/prepreparevaliditychecker.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+				submc.Self = submoduleID
+				// Load parameters from received protobuf
+				p := params.Type.(*factorypbtypes.GeneratorParams_PvcModule).PvcModule
+				// Select validity checker
+				var validityChecker PreprepareValidityChecker
+				switch ValidityCheckerType(p.ValidityChecker) {
+				case PermissiveVC:
+					validityChecker = NewPermissiveValidityChecker()
+				case CheckpointVC:
+					validityChecker = NewCheckpointValidityChecker(hashImpl, chkpVerifier, p.Membership, issParams, logger)
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/common/prepreparevaliditychecker/prepreparevaliditychecker.go]: Resolved already once I rebased in this PR.

File: protos/ordererpb/pprepvalidatorpb/pprepvalidatorpb.proto [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  oneof type {
+    contextstorepb.Origin context_store = 2;
+    dslpb.Origin          dsl           = 3;
+  }
+}
+
+message PPrepValidator {
+  option (mir.struct) = true;
+
+  uint64      validity_checker    = 1;
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply protos/ordererpb/pprepvalidatorpb/pprepvalidatorpb.proto]: Done, thanks.

File: pkg/iss/iss.go [hunk 10]
```diff
@@ -652,7 +675,7 @@ func (iss *ISS) initOrderers() error {
 				seg,
 				iss.moduleConfig.Availability.Then(t.ModuleID(fmt.Sprintf("%v", iss.epoch.Nr()))),
 				iss.epoch.Nr(),
-				orderers.PermissiveValidityChecker,
+				iss.moduleConfig.PPrepValidator.Then("permissive"),
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/iss/iss.go:677]: This is irrelevant once previous comment resolved

File: pkg/iss/iss.go [hunk 11]
```diff
@@ -222,6 +224,17 @@ func New(
...
+		//Instantiate the stateless PPV module for the permissive validity checker
+		// Instantiate a new PBFT orderer.
+		factorypbdsl.NewModule(iss.m,
+			iss.moduleConfig.PPrepValidator,
+			iss.moduleConfig.PPrepValidator.Then("permissive"),
+			tt.RetentionIndex(1<<64-1), // prevent garbage collection
+			ppv.InstanceParams(
+				ppv.PermissivePPV,
+				iss.Params.InitialMembership, // this parameter is actually not necessary for the permissiveVC
+			))
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/iss/iss.go:235]: Done :)

File: pkg/orderers/factory.go [hunk 12]
```diff
@@ -58,17 +48,11 @@ func Factory(
...
 
 				// Select validity checker
-				var validityChecker common.ValidityChecker
-				switch ValidityCheckerType(p.ValidityChecker) {
-				case PermissiveValidityChecker:
-					validityChecker = common.NewPermissiveValidityChecker()
-				case CheckpointValidityChecker:
-					validityChecker = common.NewCheckpointValidityChecker(hashImpl, chkpVerifier, segment.Membership, issParams.ConfigOffset, logger)
 
+				if p.PpvModuleId == "prepreparevaliditycheckerpermissive" { // this must change but that's the scope of a different PR
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/factory.go]: You are absolutely right, I slipped that one, changed back to chkp validator. :)

[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/factory.go]: Thanks for the catch!!

**Comment by ranchalp** [comment]
Changes addressed! :)

**Comment by matejpavlovic** [comment]
Yes, I agree, that's what we should do eventually. I think that the checkpoint validity checker (#493) is good as it is, as it's a conceptually a different thing than validating propsals. The fact that some proposals happen to contain a checkpoint is rather coincidental. The one concrete proposal validity checker that happens to validate checkpoints might, of course, "outsource" part of the validation to the checkpoint validator (e.g., deserialize the checkpoint from the raw data, have that validated by another module, and then respond to the ordering protocol), but otherwise I think that the concepts should not mix.

**Review by matejpavlovic — APPROVED** [review]
Great job Alex!
I added one more commit with a little bit of minor polishing and a few comments.
If you agree with those changes, go ahead and merge! 🚀

**Pull request merged**